src/opf/ed.jl

```diff
@@ -0,0 +1,467 @@
+    model[:ptdf_flow] = Vector{JuMP.ConstraintRef}(undef, E)
+    model.ext[:tracked_branches] = zeros(Bool, E)
+    model.ext[:ptdf_iterations] = 0
+
+    # Objective
+
+    c = [ref[:gen][g]["cost"][2] for g in 1:G]
+    q = [get!(ref[:gen][g], "reserve_cost", zeros(T, 3))[2] for g in 1:G]
+    
+    l, u = extrema(gen["cost"][1] for (i, gen) in ref[:gen])
+    (l == u == 0.0) || @warn "Data $(data["name"]) has quadratic cost terms; those terms are being ignored"
+    
+    l, u = extrema(gen["reserve_cost"][1] for (i, gen) in ref[:gen])
+    (l == u == 0.0) || @warn "Data $(data["name"]) has quadratic reserve cost terms; those terms are being ignored"
+
+    JuMP.@objective(model, Min,
+        sum(c[g]*pg[g] + ref[:gen][g]["cost"][3] for g in 1:G) +
+        sum(q[g]*r[g] + ref[:gen][g]["reserve_cost"][3] for g in 1:G) +
+        power_balance_penalty * δpb_surplus +
+        power_balance_penalty * δpb_shortage +
+        reserve_shortage_penalty * δr_shortage +
+        thermal_penalty * sum(δf)
+    )
+
+    return OPFModel{EconomicDispatch}(data, model)
+end
+
+function update!(opf::OPFModel{EconomicDispatch}, data::Dict{String,Any})
+    PM.standardize_cost_terms!(data, order=2)
+    PM.calc_thermal_limits!(data)
+    ref = PM.build_ref(data)[:it][:pm][:nw][0]
+
+    opf.data = data
+
+    L = length(ref[:load])
+    PD = sum(ref[:load][l]["pd"] for l in 1:L)
+
+    JuMP.set_normalized_rhs(opf.model[:power_balance], PD)
+
+    MRR = get(data, "minimum_reserve", 0.0)
+    JuMP.set_normalized_rhs(opf.model[:reserve_requirement], MRR)
+
+    G = length(ref[:gen])
+    rmin = (MRR > 0.0) ? [ref[:gen][g]["rmin"] for g in 1:G] : zeros(typeof(opf.model).parameters[1], G)
+    rmax = (MRR > 0.0) ? [ref[:gen][g]["rmax"] for g in 1:G] : zeros(typeof(opf.model).parameters[1], G)
+
+    JuMP.set_lower_bound.(opf.model[:r], rmin)
+    JuMP.set_upper_bound.(opf.model[:r], rmax)
+    
+    return nothing
+end
+
+function solve!(opf::OPFModel{EconomicDispatch})
+    data = opf.data
+    model = opf.model
+    T = typeof(model).parameters[1]
+    tol = model.ext[:solve_metadata][:iterative_ptdf_tol]
+
+    N = length(data["bus"])
+    G = length(data["gen"])
+    E = length(data["branch"])
+    L = length(data["load"])
+    Al = sparse(
+        [data["load"]["$l"]["load_bus"] for l in 1:L],
+        [l for l in 1:L],
+        ones(T, L),
+        N,
+        L,
+    )
+
+    Ag = sparse(
+        [data["gen"]["$g"]["gen_bus"] for g in 1:G],
+        [g for g in 1:G],
+        ones(T, G),
+        N,
+        G,
+    )
+
+    pd = [data["load"]["$l"]["pd"] for l in 1:L]
+    rate_a = [data["branch"]["$e"]["rate_a"] for e in 1:E]
+
+    p_expr = Ag * model[:pg] - Al * pd
+
+    if !model.ext[:solve_metadata][:iterative_ptdf]
+        # add entire PTDF
+        @constraint(model, model[:pf] .== model.ext[:PTDF] * p_expr)
+        optimize!(opf.model, _differentiation_backend = MathOptSymbolicAD.DefaultBackend())
+        return
+    end
```

Comment:
Would we every run a PTDF-based formulation without an iterative solve?
I am OK not supporting this, and recommend the phase-angle formulation should someone wants to solve the whole problem in one go.

Reply:
Probably not, but it does help in case we need to debug the iterative solver. Shall I remove it?

Reply:
> Note if keeping it, still needs an update to store the entire PTDF in `:ptdf_flow`

I am not sure I fully understand what you mean here. What is "update" referring to?

Reply:
Right now it is an unnamed constraint which will cause issues when extracting the duals. It should be stored in the same container as the iterative PTDF uses, and `tracked_branches` should be set accordingly as well. 

Or we can remove support for non-iterative solving.

Reply:
> Probably not, but it does help in case we need to debug the iterative solver. Shall I remove it?

TBH it's more the flow of the code that feels a bit weird to me. There's this little block that shunts all the iterative algorithm.

An alternative would be to add all the PTDF constraints, then run the same iterative algorithm. It will converge in a single iteration, and you can use the same log system, solution querying, etc... After the first call to `optimize!`, no PTDF constraint will be violated, and the algorithm with terminate.